### PR TITLE
Remove TH/THC Storage data and copy functions

### PR DESCRIPTION
**Pull request by kurtamohler**
Part of #67852

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

### CODE REVIEW

File: torch/csrc/generic/StorageMethods.cpp [hunk 1]
```diff
@@ -41,7 +60,7 @@ static PyObject * THPStorage_(isPinned)(PyObject *_self, PyObject *noargs)
   HANDLE_TH_ERRORS
   auto self = (THPStorage*)_self;
 #if defined(USE_CUDA)
-  return PyBool_FromLong(at::globalContext().isPinnedPtr(THWStorage_(data)(LIBRARY_STATE self->cdata)));
+  return PyBool_FromLong(at::globalContext().isPinnedPtr(reinterpret_cast<c10::StorageImpl*>(self->cdata)->data<scalar_t>()));
```
Comment (peterbell10) [inline torch/csrc/generic/StorageMethods.cpp:63]: Why do you need `reinterpret_cast` for `cdata` but not other uses of `THWStorage`?

File: torch/csrc/generic/StorageSharing.cpp [hunk 2]
```diff
@@ -160,7 +167,14 @@ static PyObject * THPStorage_(shareFd)(PyObject *_self, PyObject *noargs)
   } else {
     THWStoragePtr new_storage(
         THPStorage_(newFdStorage)(storage->nbytes() / sizeof(scalar_t)));
-    THWStorage_(copy)(new_storage, storage);
+
+    at::Storage new_storage_aten(
+      c10::intrusive_ptr<c10::StorageImpl>(
+        new_storage.get(),
+        c10::raw::DontIncreaseRefcount()));
```
Comment (peterbell10) [inline torch/csrc/generic/StorageSharing.cpp:174]: I think your refcounting issues stem from here, the `intrusive_ptr` needs to increase the refount. You can do this with `ATen/core/UnsafeFromTH.h`
```suggestion
    at::Storage new_storage_aten = at::unsafeStorageFromTH(new_storage.get(), /*retain=*/true);
```

File: torch/csrc/generic/StorageMethods.cpp (hunk 1 above)
Comment (kurtamohler) [reply torch/csrc/generic/StorageMethods.cpp:63]: Right, these casts aren't needed

File: aten/src/THC/generic/THCStorage.cpp [hunk 3]
```diff
@@ -6,11 +6,6 @@
 #include <c10/util/typeid.h>
 
 
-scalar_t* THCStorage_(data)(THCState *state, const THCStorage *self)
-{
-  return self->data<scalar_t>();
-}
-
 int THCStorage_(elementSize)(THCState *state)
```
Comment (ngimel) [inline aten/src/THC/generic/THCStorage.cpp:9]: I think (elementSize) can also be removed (even though it's not the goal of this PR)

**Review by ngimel — APPROVED** [review]

File: aten/src/THC/generic/THCStorage.cpp (hunk 3 above)
Comment (kurtamohler) [reply aten/src/THC/generic/THCStorage.cpp]: Done. There may be other functions that can be easily removed too, but if so, I'll include them in future PRs. I'm splitting up the TH/THC Storage removal into several PRs

**Pull request closed without merging**